packages/access-client/src/encoding.js

```diff
@@ -13,89 +13,77 @@
  * @module
  */
 /* eslint-disable unicorn/prefer-spread */
-import { CarReader } from '@ipld/car/reader'
-import { CarWriter } from '@ipld/car/writer'
+import { CarReader } from './utils/car-reader-sync.js'
+import * as CarBufferWriter from '@ipld/car/buffer-writer'
 import { Delegation } from '@ucanto/core/delegation'
 import * as u8 from 'uint8arrays'
 // eslint-disable-next-line no-unused-vars
 import * as Types from '@ucanto/interface'
 
 /**
- * @param {AsyncIterable<Uint8Array>} iterable
- */
-function collector(iterable) {
-  const chunks = []
-  const cfn = (async () => {
-    for await (const chunk of iterable) {
-      chunks.push(chunk)
-    }
-    return u8.concat(chunks)
-  })()
-  return cfn
-}
-
-/**
+ * Encode delegations as bytes
+ *
  * @param {Types.Delegation[]} delegations
- * @param {import('uint8arrays/to-string').SupportedEncodings} encoding
  */
-export async function encodeDelegations(delegations, encoding = 'base64url') {
-  if (delegations.length === 0) {
-    return ''
+export function delegationsToBytes(delegations) {
+  if (!Array.isArray(delegations) || delegations.length === 0) {
+    throw new Error('Delegations required to be an non empty array.')
   }
 
-  const roots = delegations.map((d) => d.root.cid)
-
-  // @ts-ignore
-  const { writer, out } = CarWriter.create(roots)
-  const collection = collector(out)
+  const roots = delegations.map(
+    (d) => /** @type {CarBufferWriter.CID} */ (d.root.cid)
+  )
+  const cids = new Set()
+  /** @type {CarBufferWriter.Block[]} */
+  const blocks = []
+  let byteLength = 0
 
   for (const delegation of delegations) {
     for (const block of delegation.export()) {
-      // @ts-ignore
-      await writer.put(block)
+      const cid = block.cid.toV1().toString()
+      if (!cids.has(cid)) {
+        byteLength += CarBufferWriter.blockLength(
+          /** @type {CarBufferWriter.Block} */ (block)
+        )
+        blocks.push(/** @type {CarBufferWriter.Block} */ (block))
+        cids.add(cid)
+      }
     }
   }
-  await writer.close()
-
-  const bytes = await collection
-
-  return u8.toString(bytes, encoding)
-}
+  const headerLength = CarBufferWriter.estimateHeaderLength(roots.length)
+  const writer = CarBufferWriter.createWriter(
+    new ArrayBuffer(headerLength + byteLength),
```

Comment:
This is kinda weird seeing as we have all the blocks already. Would be nicer if it just collected all the buffers and joined them together when we call `close()`